rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+    /// transformed parameters or generated quantities.
+    ///
+    /// The parameters are returned in the order they are declared.
+    /// Multivariate parameters are return in column-major (more
+    /// generally last-index major) order.  Parameters are separated with
+    /// periods (`.`).  For example, `a[3]` is written `a.3` and `b[2,
+    /// 3]` as `b.2.3`.  The numbering follows Stan and is indexed from 1.
+    pub fn param_unc_names(&mut self) -> &str {
+        let cstr =
+            unsafe { CStr::from_ptr(self.lib.borrow().0.bs_param_unc_names(self.model.as_ptr())) };
+        cstr.to_str()
+            .expect("Stan model has invalid parameter names")
+    }
+
+    /// Number of parameters in the model on the constrained scale.
+    /// Will also count transformed parameters and generated quantities if requested
+    pub fn param_num(&self, include_tp: bool, include_gq: bool) -> usize {
+        unsafe {
+            self.lib.borrow().0.bs_param_num(
+                self.model.as_ptr(),
+                include_tp as c_int,
+                include_gq as c_int,
+            )
+        }
+        .try_into()
+        .expect("Stan returned an invalid number of parameters")
+    }
+
+    /// Return the number of parameters on the unconstrained scale.
+    /// In particular, this is the size of the slice required by the log_density functions.
+    pub fn param_unc_num(&self) -> usize {
+        unsafe { self.lib.borrow().0.bs_param_unc_num(self.model.as_ptr()) }
+            .try_into()
+            .expect("Stan returned an invalid number of parameters")
+    }
+
+    /// Compute the log of the prior times likelihood density
+    ///
+    /// Drop jacobian determinant terms if `jacobian == false` and
+    /// drop constant terms of the density if `propto == true`.
+    pub fn log_density(&self, theta_unc: &[f64], propto: bool, jacobian: bool) -> Result<f64> {
+        let n = self.param_unc_num();
+        assert_eq!(
+            theta_unc.len(),
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        let mut val = 0.0;
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_log_density(
+                self.model.as_ptr(),
+                propto as c_int,
+                jacobian as c_int,
+                theta_unc.as_ptr(),
+                &mut val,
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(val)
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Compute the log of the prior times likelihood density and its gradient
+    ///
+    /// Drop jacobian determinant terms if `jacobian == false` and
+    /// drop constant terms of the density if `propto == true`.
+    /// The gradient of the log density is stored in `grad`.
+    pub fn log_density_gradient(
```

Comment:
For this and the hessian function, do you think it is valuable (or idiomatic?) to provide versions which allocate memory for the output for you? We do in Julia/Python, but that's following established conventions in those two languages

Reply:
I think it is good to have the non-allocating functions around. I can add additional functions that return new arrays, but I don't really think that's necessary. Rust is generally sufficiently low-level that I think it wouldn't be unreasonable to expect the user to do that.

Reply:
Yeah, we definitely want the non-allocating ones! I was more wondering if you thought it was worth adding some which return new arrays in addition to these, but I agree they aren't necessary.

If someone asks for them, we could always add new functions which do that later on.